dynamiqs/utils/utils.py

```diff
@@ -316,26 +337,28 @@ def ptrace(x: Tensor, keep: int | tuple[int, ...], dims: tuple[int, ...]) -> Ten
 
     # trace out x over unkept dimensions
     batch_dims = x.shape[:-2]
-    if is_ket(x):
+    if is_ket(x) or is_bra(x):
         x = x.view(-1, *dims)  # e.g. (..., 20, 2, 5)
         eq = ''.join(['...'] + eq1 + [',...'] + eq2)  # e.g. '...abc,...ade'
         x = torch.einsum(eq, x, x.conj())  # e.g. (..., 2, 5, 2, 5)
-    else:
+    elif is_dm(x):
         x = x.view(-1, *dims, *dims)  # e.g. (..., 20, 2, 5, 20, 2, 5)
         eq = ''.join(['...'] + eq1 + eq2)  # e.g. '...abcade'
         x = torch.einsum(eq, x)  # e.g. (..., 2, 5, 2, 5)
+    else:
+        raise TypeError('Input tensor is not a ket, bra or density matrix.')
```

Comment:
```suggestion
        raise ValueError('Input tensor is not a ket, bra or density matrix.')
```

Reply:
Need to add a `Raises` statement in the docstring (here and in the other functions).